src/sentry/utils/services.py

```diff
@@ -235,32 +238,67 @@ def execute(*args, **kwargs):
+
+            # Enqueue all of the secondary backend requests first since these
+            # are non-blocking queue insertions. (Since the primary backend
+            # executor queue insertion can block, if that queue was full the
+            # secondary requests would have to wait unnecessarily to be queued
+            # until the after the primary request can be enqueued.)
+            # NOTE: If the same backend is both the primary backend *and* in
+            # the secondary backend list -- this is unlikely, but possible --
+            # this means that one of the secondary requests will be queued and
+            # executed before the primary request is queued.  This is such a
+            # strange usage pattern that I don't think it's worth optimizing
+            # for.)
+            results = [None] * len(selected_backend_names)
+            for i, backend_name in enumerate(selected_backend_names[1:], 1):
                 try:
                     backend, executor = self.__backends[backend_name]
                 except KeyError:
-                    if is_primary:
-                        raise self.InvalidBackend(
-                            '{!r} is not a registered backend.'.format(backend_name))
-                    else:
-                        logger.warning(
-                            '%r is not a registered backend and will be ignored.',
-                            backend_name,
-                            exc_info=True)
-                        continue
-
-                results[backend_name] = executor.submit(
-                    functools.partial(getattr(backend, attribute_name), *args, **kwargs),
-                    block=is_primary,
-                )
+                    logger.warning(
+                        '%r is not a registered backend and will be ignored.',
+                        backend_name,
+                        exc_info=True)
+                else:
+                    results[i] = executor.submit(
+                        functools.partial(call_backend_method, backend),
+                        priority=1,
+                        block=False,
+                    )
+
+            # The primary backend is scheduled last since it may block the
+            # calling thread. (We don't have to protect this from ``KeyError``
+            # since we already ensured that the primary backend exists.)
+            backend, executor = self.__backends[selected_backend_names[0]]
+            results[0] = executor.submit(
+                functools.partial(call_backend_method, backend),
+                priority=0,
+                block=True,
+            )
 
             if self.__callback_func is not None:
-                FutureSet(results.values()).add_done_callback(
-                    lambda *a, **k: self.__callback_func(attribute_name, callargs, results)
+                FutureSet(filter(None, results)).add_done_callback(
```

Comment:
This needs to be filtered in case we ignored any bad references in the secondary backend list.